### PR TITLE
BeamCsvScenarioWriter compatible with ScenarioInfo classes

### LINKED ISSUE #2894: why is writing not possible directly after reading something with BeamCsvScenarioReader symmetrically?

### PR DESCRIPTION
**Pull request by carloscaldas**
Resolves #2894 

This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2961)

Example of usage below:
```Scala
// Implicits supports conversion from Iterable and also Array
import beam.utils.csv.readers.BeamCsvScenarioReader
import beam.utils.csv.writers.BeamCsvScenarioWriter._

val vehicles: Iterable[VehicleInfo] =
  BeamCsvScenarioReader.readVehiclesFile("vehicles.csv")
vehicles.writeCsvVehiclesFile("vehicles_output.csv")

val households: Array[HouseholdInfo] =
  BeamCsvScenarioReader.readHouseholdsFile("householdsFile.csv", vehicles)
vehicles.writeCsvVehiclesFile("households_output.csv")

val persons: Array[PersonInfo] =
  BeamCsvScenarioReader.readPersonsFile("persons.csv")
persons.writeCsvPersonsFile("persons_out.csv")
```

### CODE REVIEW

File: src/main/scala/beam/utils/csv/writers/PopulationCsvWriter.scala [hunk 1]
```diff
@@ -36,41 +40,76 @@ object PopulationCsvWriter extends ScenarioCsvWriter {
...
+        HouseholdId(houseHoldId),
+        Try(rank.toInt).getOrElse(0),
+        Try(personAge.toInt).getOrElse(0),
+        isFemale = isFemale,
+        valueOfTime = Try(valueOfTime.toString.toDouble).getOrElse(0),
+        excludedModes = excludedModes
       )
-      values.mkString("", FieldSeparator, LineSeparator)
     }
+    contentIterator(values)
```
[the remark this answers is not in the document]
Comment (JustinPihony) [reply src/main/scala/beam/utils/csv/writers/PopulationCsvWriter.scala:85]: Although this one seems more pertinent - so maybe keep as is...but I'd argue not still due to unnecessary object allocation and traversal, when you only need `toLine`

File: src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package beam.utils.csv.writers
+
+import scala.language.implicitConversions
+
+import beam.utils.scenario.{HouseholdInfo, PersonInfo, VehicleInfo}
+import com.typesafe.scalalogging.LazyLogging
+
+object BeamCsvScenarioWriter extends LazyLogging {
```
Comment (REASY) [inline src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala:8]: What are all these implicits about? Could you, please, provide an example of the usage new writers?

File: src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      arr
+    }
+  }
+
+  implicit class VehicleWriter(arr: Array[VehicleInfo]) {
+    def this(elements: Iterable[VehicleInfo]) {
+      this(elements.toArray)
+    }
+
+    def writeCsvHouseholdsFile(outputFileStr: String): Array[VehicleInfo] = {
```
Comment (REASY) [inline src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala:26]: typo? `VehicleWriter` but writeCsv**Households**File

**Review by REASY — CHANGES_REQUESTED** [review]
The branch is not compiling: https://beam-ci.tk/job/beam-4ci/8982/consoleFull, please, fix it

File: src/main/scala/beam/utils/csv/writers/HouseholdsCsvWriter.scala [hunk 4]
```diff
@@ -28,15 +30,33 @@ object HouseholdsCsvWriter extends ScenarioCsvWriter with StrictLogging {
...
-        incomeValue = h.getIncome.getIncome,
-        locationX = attributes.getAttribute(id, "homecoordx").toString,
-        locationY = attributes.getAttribute(id, "homecoordy").toString
-      ).toString
+      HouseholdInfo(
+        householdId = HouseholdId(id),
+        income = h.getIncome.getIncome,
+        locationX = Try(attributes.getAttribute(id, "homecoordx").toString.toDouble).getOrElse(0),
+        locationY = Try(attributes.getAttribute(id, "homecoordy").toString.toDouble).getOrElse(0),
+        // TODO: this information is not available in CSV file
```
[the remark this answers is not in the document]
Comment (carloscaldas) [reply src/main/scala/beam/utils/csv/writers/HouseholdsCsvWriter.scala:40]: Totally agree. Issue created and linked in the TODO

File: src/main/scala/beam/utils/csv/writers/PopulationCsvWriter.scala [hunk 5]
```diff
@@ -36,41 +40,76 @@ object PopulationCsvWriter extends ScenarioCsvWriter {
...
+      val rank = String.valueOf(personAttributes.getAttribute(personId.toString, "rank"))
+
+      val houseHoldId: String = personIdToHouseHoldId.get(personId).map(_.toString).getOrElse("")
+
+      PersonInfo(
+        PersonId(personId.toString),
+        HouseholdId(houseHoldId),
+        Try(rank.toInt).getOrElse(0),
+        Try(personAge.toInt).getOrElse(0),
+        isFemale = isFemale,
```
[the remark this answers is not in the document]
Comment (carloscaldas) [reply src/main/scala/beam/utils/csv/writers/PopulationCsvWriter.scala:80]: In this case is the opposite. I need the parameter name everywhere. The variable having the same name does not mean is placed in the right place.

File: src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala (hunk 3 above)
Comment (carloscaldas) [reply src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala]: nice catch. thx

File: src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala (hunk 2 above)
Comment (carloscaldas) [reply src/main/scala/beam/utils/csv/writers/BeamCsvScenarioWriter.scala]: Added to the PR description and also added as comments to the file

File: src/main/scala/beam/utils/csv/writers/PopulationCsvWriter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (JustinPihony) [reply src/main/scala/beam/utils/csv/writers/PopulationCsvWriter.scala:81]: Yah, wasn't sure since I didn't use an IDE to check - makes sense.

**Review by JustinPihony — APPROVED** [review]

**Comment by JustinPihony** [comment]
There is a wrong toString somewhere, as the tests fail due to toString'ing an Array - maybe add explicit types to double check you have what you want?

> 13:41:09 ModeExclusionTest:
> 13:41:09 Exclusion of mode
> 13:41:09 - should check for removal of single mode *** FAILED *** (60 milliseconds)
> 13:41:09   Array("WrappedArray(car)") was not equal to Array("car") (ModeExclusionTest.scala:37)
> 13:41:09 - should check for removal of multiple mode *** FAILED *** (0 milliseconds)
> 13:41:09   Array("WrappedArray(car", " walk)") was not equal to Array("car", "walk") (ModeExclusionTest.scala:41)
> 13:41:09 - should check for removal of empty mode *** FAILED *** (1 millisecond)
> 13:41:09   Array("WrappedArray()") was not equal to Array() (ModeExclusionTest.scala:45)

**Comment by carloscaldas** [comment]
Test now!

**Review by REASY — APPROVED** [review]

**Comment by carloscaldas** [comment]
Test now!

**Comment by carloscaldas** [comment]
Test now!

File: src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala [hunk 6]
```diff
@@ -36,16 +37,13 @@ object AvailableModeUtils extends LazyLogging {
...
-      case Some(modes) =>
-        if (modes.isEmpty) {
-          Array.empty[String]
-        } else {
-          modes.split(",")
-        }
-      case None => Array.empty[String]
+    )
+    maybeExcludedModes match {
+      case Some(modes: Iterable[_])   => modes.map(_.toString).filterNot(v => isBlank(v)).toArray
```
Comment (JustinPihony) [inline src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala:44]: The split was lost here - it seems it should still be needed?

Comment (JustinPihony) [reply src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala:44]: Could be the reason for two new failures:

> 13:24:42 CarSharingSpec:
> 13:24:42 Running a car-sharing-only scenario with abundant cars
> 13:24:42 - must result in everybody driving *** FAILED *** (18 seconds, 273 milliseconds)
> 13:24:42   4 did not equal 0 Someone wasn't driving even though everybody wants to and cars abound. (CarSharingSpec.scala:147)
> 13:24:42 Running a car-sharing-only scenario with one car per person at home
> 13:24:42 - must result in everybody driving *** FAILED *** (21 seconds, 21 milliseconds)
> 13:24:42   82 did not equal 0 Someone wasn't driving even though everybody wants to and cars abound. (CarSharingSpec.scala:147)

Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala:44]: I put some log with the previous output and the new output. with the current change the output is the same

File: src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala [hunk 7]
```diff
@@ -36,16 +37,15 @@ object AvailableModeUtils extends LazyLogging {
...
-        } else {
-          modes.split(",")
-        }
-      case None => Array.empty[String]
+    )
+    maybeExcludedModes.map(_.toString)
+    maybeExcludedModes match {
+      case Some(modes: Array[String]) => modes.filterNot(isBlank)
+      case Some(modes: Iterable[_])   => modes.flatMap(_.toString.split(",")).filterNot(isBlank).toArray
+      case Some(modes)                => modes.toString.split(",").filterNot(isBlank)
```
Comment (JustinPihony) [inline src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala:47]: After adding the scenario to handle strings it is now failing on 

> 20:06:10 ParkingSpec:
> 20:06:10 Parking system
...
> 20:06:10 - must very expensive parking should reduce driving *** FAILED *** (2 minutes, 5 seconds)
> 20:06:10   8 was not greater than 20 (ParkingSpec.scala:225)
> 20:06:10 - must no parking stalls should reduce driving *** FAILED *** (1 minute, 34 seconds)
> 20:06:10   8 was not greater than 20 (ParkingSpec.scala:239)

I'm wondering if this splitting of string (which in some parts was new IIRC) results in test breakage that was never correct?

Comment (JustinPihony) [reply src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala:47]: Although, I just re-ran locally and it worked...I am going to kick off again and see if it is a flaky test...I think I may have seen this earlier also...

Comment (JustinPihony) [reply src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala:47]: Yah - Already created as #2957

**Pull request merged**